src/neurostatslib/utils.py

```diff
@@ -4,44 +4,243 @@
+
+    Check that all arrays in iterable has the expected dimensionality.
+
+    Parameters
+    ----------
+    iterable :
+    Array-like object containing numpy or jax.numpy NDArrays.
+    expected_dim :
+    Number of expected dimension for the NDArrays.
+
+    Returns
+    -------
+    True if all the arrays has the expected number of dimension, False otherwise.
+    """
+    return not any(array.ndim != expected_dim for array in iterable)
+
+
+def convolve_1d_trials(
+    basis_matrix: ArrayLike,
+    time_series: Union[Iterable[NDArray], NDArray, Iterable[jnp.ndarray], jnp.ndarray],
+) -> List[jnp.ndarray]:
+    """
+    Convolve trial time series with a basis matrix.
+
+    This function checks if all trials have the same duration. If they do, it uses a fast method
+    to convolve all trials with the basis matrix at once. If they do not, it falls back to convolving
+    each trial individually.
+
+    Parameters
+    ----------
+    basis_matrix :
+        The basis matrix with which to convolve the trials. Shape (n_basis_funcs, window_size).
+    time_series :
+        The time series of trials to convolve with the basis matrix. It should be a list of 2D arrays,
+        where each array represents a trial and its second dimension matches the first dimension
+        of the basis_matrix. Each trial has shape (n_neurons, n_timebins_trial).
+
+    Returns
+    -------
+    :
+        The convolved trials. It is a list of 3D arrays, where each array represents a convolved trial.
+        Each element of the list will have shape (n_neurons, n_basis_funcs, n_timebins_trial - window_size - 1).
+
+    Raises
+    ------
+    ValueError
+        - If basis_matrix is not a 2D array-like object.
+        - If trials_time_series is not an iterable of 2D array-like objects.
+        - If trials_time_series contains empty trials.
+        - If the number of time points in each trial is less than the window size.
+    """
+
+    basis_matrix = jnp.asarray(basis_matrix)
+    # check input size
+    if basis_matrix.ndim != 2:
+        raise ValueError("basis_matrix must be a 2 dimensional array-like object.")
+
+    try:
+        # this should fail for variable trial length
+        time_series = jnp.asarray(time_series)
+        if time_series.ndim != 3:
+            raise ValueError
+
+    except ValueError:
+        # convert each trial to array
+        time_series = [jnp.asarray(trial) for trial in time_series]
+        if not check_dimensionality(time_series, 2):
+            raise ValueError(
+                "trials_time_series must be an iterable of 2 dimensional array-like objects."
+            )
+
+    if any(k == 0 for trial in time_series for k in trial.shape) | (
+        len(time_series) == 0
+    ):
+        raise ValueError("trials_time_series should not contain empty trials!")
+
+    # Check window size
+    ws = len(basis_matrix[0])
+    if any(trial.shape[1] < ws for trial in time_series):
+        raise ValueError(
```

Comment:
Should it raise a warning or an error. Could it be possible to just remove the trial in this case? 
In pynapple,  there is drop_short_intervals to solve this issue before calling the GLM. But I was wondering if it should just go on especially if it's only a small percentage of intervals that are too shorts.